ocaml/libs/tracing/tracing.ml

```diff
@@ -573,7 +575,7 @@ let set ?enabled ?attributes ?endpoints ~uuid () =
         Hashtbl.clear Spans.spans ;
         Hashtbl.clear Spans.finished_spans
     )
-  ) else
+  ) else if not (get_observe ()) then
     set_observe true
```

Comment:
The set call here can race with another set call from the API. It is better to do the (atomic) set operation while you are holding the lock.

Reply:
In fact 'get_tracer_providers' above is not thread-safe: it uses a hashtable, and iterates on it without holding any locks.